tapdb/universe_federation.go

```diff
@@ -174,4 +214,148 @@ func (u *UniverseFederationDB) LogNewSyncs(ctx context.Context,
+		// Upsert general federation sync config.
+		err := db.SetFederationGlobalSyncConfig(
+			ctx, globalSyncConfig.ProofTypes.String(),
+		)
+		if err != nil {
+			return err
+		}
+
+		// Upsert universe specific sync configs.
+		for i := range uniSyncConfigs {
+			var (
+				config = uniSyncConfigs[i]
+
+				uniID        = config.UniverseID
+				groupPubKey  []byte
+				assetIDBytes []byte
+			)
+
+			// If the group key is set, then we'll serialize it
+			// into bytes.
+			if uniID.GroupKey != nil {
+				groupPubKey = uniID.GroupKey.SerializeCompressed()
+			} else {
+				// If the group key is not set, then we'll use
+				// the asset ID. The group key supersedes the
+				// asset ID.
+				assetIDBytes = uniID.AssetID[:]
+			}
+
+			params := UpsertFedUniSyncConfigParams{
+				AssetID:   assetIDBytes,
+				GroupKey:  groupPubKey,
+				ProofType: config.ProofTypes.String(),
+			}
+			err := db.UpsertFederationUniSyncConfig(ctx, params)
+			if err != nil {
+				return err
+			}
+		}
+
+		return nil
+	})
+}
+
+// QueryFederationSyncConfigs returns the general and universe specific
+// federation sync configs.
+func (u *UniverseFederationDB) QueryFederationSyncConfigs(
+	ctx context.Context) (*universe.FedGlobalSyncConfig,
+	[]*universe.FedUniSyncConfig, error) {
+
+	var (
+		readTx UniverseFederationOptions
+
+		globalConfig *universe.FedGlobalSyncConfig
+		uniConfigs   []*universe.FedUniSyncConfig
+	)
+
+	err := u.db.ExecTx(ctx, &readTx, func(db UniverseServerStore) error {
+		// Query for global sync config.
+		proofTypesStr, err := db.QueryFederationGlobalSyncConfig(
+			ctx,
+		)
+		if err != nil {
+			return err
+		}
+		proofTypes, err := universe.ParseProofType(
+			proofTypesStr,
+		)
+		if err != nil {
+			return err
+		}
+
+		globalConfig = &universe.FedGlobalSyncConfig{
+			ProofTypes: proofTypes,
+		}
+
+		// If no rows in result set, then we haven't set the config
+		// yet, so we'll default to true.
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil
+		}
+
+		// Query for universe specific sync configs.
+		configsDb, err := db.QueryFederationUniSyncConfigs(ctx)
+		if err != nil {
+			return err
+		}
+
+		for i := range configsDb {
+			conf := configsDb[i]
+
+			proofTypes, err := universe.ParseProofType(
+				conf.ProofType,
+			)
+			if err != nil {
+				return err
+			}
+
+			// Construct group key public key from bytes.
+			var pubKey *btcec.PublicKey
+			if conf.GroupKey != nil {
+				pubKey, err = btcec.ParsePubKey(conf.GroupKey)
+				if err != nil {
+					return fmt.Errorf("unable to parse "+
+						"group key: %v", err)
+				}
+			}
+
+			// Construct asset ID from bytes.
+			var assetID asset.ID
+			copy(assetID[:], conf.AssetID)
+
+			uniID := universe.Identifier{
+				AssetID:  assetID,
+				GroupKey: pubKey,
+			}
+
+			uniConfig := universe.FedUniSyncConfig{
+				UniverseID: uniID,
+				ProofTypes: proofTypes,
+			}
+			uniConfigs = append(uniConfigs, &uniConfig)
```

Comment:
Should we pre-allocate the `uniConfigs` right before the `for` loop and then use index based insertion, since we know how many elements we expect?